File: src/core/brain.js
(function (F) {
    'use strict';
    const { clamp, wrap, TAU } = F;
    /* Hybrid demonstrator, NOT a fitted biological circuit or a connectome-only
     * emulation. All state origins and controller assumptions are disclosed.
     * Critically: step() has no access to body/world ground truth. */
    class NeuralController {
        constructor(circuit, seed) { this.circuit = circuit; this.rng = new F.RNG(seed ^ 0x7cab1); this.a = new Float64Array(circuit.N); this.a.fill(.05); this.adapt = new Float64Array(circuit.N); this.lesions = new Set(); this.stim = new Map(); this.phaseEstimate = 0; this.heading = 0; this.confidence = 0; this.goal = 1.0; this.clock = 0; this.nextGoal = 9; this.avoidSide = 1; this.prevContact = 0; this.odorMode = 'explore'; this.turnMemory = 0; this.contactTime = 0; this.recoveryTime = 0; this.output = { yawRate: 0, forwardSpeed: 0, verticalSpeed: 0 }; this.lastCue = null; this.error = 0; this.recNorm = 0; }
        val(type, k = 0) { return this.a[this.circuit.pop[type][k]] || 0; }
        mean(type) { let is = this.circuit.pop[type]; return is.reduce((s, i) => s + this.a[i], 0) / is.length; }
        setTarget(i, value, dt, tau = .13) { if (this.lesions.has(i)) {
            this.a[i] = 0;
            return;
        } const stim = this.stim.get(i); const extra = stim ? stim.amplitude : 0; this.a[i] += (clamp(value + extra) - this.a[i]) * (1 - Math.exp(-dt / tau)); }
        step(s, dt, config) {
            this.clock += dt;
            const c = this.circuit, old = this.a.slice();
            // Decoding of the far-field cue occurs inside the brain from 64 observations.
            let cx = 0, cy = 0, cs = 0;
            for (let i = 0; i < 64; i++) {
                const p = i / 64 * TAU - Math.PI;
                cx += s.panorama[i] * Math.cos(p);
                cy += s.panorama[i] * Math.sin(p);
                cs += s.panorama[i];
            }
            const visible = cs > .8;
            this.lastCue = visible ? wrap(-Math.atan2(cy, cx)) : null;
            // Explicit angular-observer scaffold (HD); no direct assignment from true yaw.
            this.phaseEstimate = wrap(this.phaseEstimate + s.angularVelocity * dt);
            if (visible)
                this.phaseEstimate = wrap(this.phaseEstimate + wrap(this.lastCue - this.phaseEstimate) * (1 - Math.exp(-dt / .24)));
            for (const i of c.pop.HD)
                this.setTarget(i, .05 + .90 * Math.exp(3.4 * (Math.cos(c.nodes[i].angle - this.phaseEstimate) - 1)), dt, .07);
            let hx = 0, hy = 0;
            for (const i of c.pop.HD) {
                const a = Math.max(0, this.a[i] - .04);
                hx += a * Math.cos(c.nodes[i].angle);
                hy += a * Math.sin(c.nodes[i].angle);
            }
            const scaffold = Math.atan2(hy, hx);
            const rec = c.propagate(old, config.graphGain);
            this.recNorm = rec.reduce((x, y) => x + y, 0) / c.realCount;
            for (let i = 0; i < c.realCount; i++) {
                const n = c.nodes[i], bump = Math.exp(3 * (Math.cos(n.angle - scaffold) - 1));
                const omega = clamp(s.angularVelocity / 2, -1, 1);
                let target = .055 + .77 * bump;
                if (n.type === 'PENa' || n.type === 'PENb')
                    target = .07 + .58 * bump + .18 * Math.max(0, omega * (n.hemisphere === 'R' ? 1 : -1));
                if (n.type === 'PEG')
                    target = .07 + .66 * bump;
                target += .30 * rec[i] - .045;
                this.adapt[i] += (old[i] - this.adapt[i]) * dt / 2;
                target -= .025 * this.adapt[i];
                this.setTarget(i, target, dt, n.type === 'PEG' ? .28 : .11);
            }
            // Equalize cell counts by anatomical column before decoding EPG heading.
            const bins = new Float64Array(8), counts = new Uint16Array(8);
            for (const i of c.pop.EPG) {
                const b = (c.nodes[i].indexFix - 1) % 8;
                bins[b] += Math.max(0, this.a[i] - .025);
                counts[b]++;
            }
            let ex = 0, ey = 0, total = 0;
            for (let k = 0; k < 8; k++) {
                bins[k] /= Math.max(1, counts[k]);
                ex += bins[k] * Math.cos(k * TAU / 8);
                ey += bins[k] * Math.sin(k * TAU / 8);
                total += bins[k];
            }
            this.confidence = total > .01 ? Math.hypot(ex, ey) / total : 0;
            if (total > .01)
                this.heading = Math.atan2(ey, ex);
            this.headingBins = Array.from(bins);
            if (!visible && this.confidence > .1)
                this.phaseEstimate = wrap(this.phaseEstimate + wrap(this.heading - this.phaseEstimate) * dt * .08);
            let dangerL = Math.max(...s.nearRanges.slice(0, 4).map(r => Math.exp(-r / 2.2))), dangerR = Math.max(...s.nearRanges.slice(5).map(r => Math.exp(-r / 2.2))), front = Math.exp(-s.nearRanges[4] / 2.2);
            const clearance = Math.min(...s.nearRanges.slice(3, 6));
            if (this.turnMemory) {
                const direction = this.turnMemory > 0 ? 1 : -1;
                const remaining = clamp(Math.abs(this.turnMemory) - direction * s.angularVelocity * dt, .001, TAU);
                this.turnMemory = direction * remaining;
                if (remaining <= .05 && clearance > 4.5 && !s.contact) this.turnMemory = 0;
            } else if (clearance < 3.5 || s.contact) {
                if (Math.abs(dangerL - dangerR) > .08) this.avoidSide = dangerL > dangerR ? 1 : -1;
                this.turnMemory = this.avoidSide * 1.3;
            }
            this.prevContact = s.contact;
            const avoiding = !!this.turnMemory;
            this.recoveryTime = Math.max(0, this.recoveryTime - dt);
            this.contactTime = clamp(this.contactTime + (s.contact && avoiding ? dt : -dt * .5), 0, 1);
            if (avoiding && !this.recoveryTime && this.contactTime > .3) {
                this.recoveryTime = .4;
                this.contactTime = 0;
            }
            this.setTarget(c.pop.SENSE[0], dangerL, dt, .08);
            this.setTarget(c.pop.SENSE[1], dangerR, dt, .08);
            this.setTarget(c.pop.SENSE[2], s.odor[0], dt, .15);
            this.setTarget(c.pop.SENSE[3], s.odor[1], dt, .15);
            if (config.task === 'heading') {
                this.goal = config.goalAngle;
                this.odorMode = 'heading';
            }
            else {
                if (this.clock > this.nextGoal) {
                    this.goal = wrap(this.goal + (.5 + this.rng.next()) * this.rng.signed() * 2);
                    this.nextGoal = this.clock + 7 + this.rng.next() * 7;
                }
                if (Math.max(...s.odor) > .10) {
                    const odorDiff = this.val('SENSE', 3) - this.val('SENSE', 2);
                    this.goal = wrap(this.heading + clamp(odorDiff * 9, -1, 1) + (s.odorChange < -.02 ? .50 : 0));
                    this.odorMode = 'chemotaxis';
                }
                else
                    this.odorMode = 'explore';
                if (Math.max(...s.odor) > .88) {
                    this.goal = wrap(this.heading + .7);
                    this.odorMode = 'dwell';
                }
            }
            for (const i of c.pop.GOAL)
                this.setTarget(i, .02 + .94 * Math.exp(3.3 * (Math.cos(c.nodes[i].angle - this.goal) - 1)), dt, .2);
            let gx = 0, gy = 0;
            for (const i of c.pop.GOAL) {
                gx += this.a[i] * Math.cos(c.nodes[i].angle);
                gy += this.a[i] * Math.sin(c.nodes[i].angle);
            }
            const goalRead = Math.atan2(gy, gx);
            this.error = wrap(goalRead - this.heading);
            let avoid = clamp(this.val('SENSE', 0) - this.val('SENSE', 1) + Math.max(0, s.danger - .25) * this.avoidSide, -1, 1), e = this.error;
            if (avoiding) avoid = .85 * (this.turnMemory > 0 ? 1 : -1);
            // PFL3-inspired *synthetic* directional readout; not claimed as anatomical PFL3.
            for (let k = 0; k < 16; k++) {
                const attention = avoiding ? 1 : .30 + .70 * this.a[c.pop.GOAL[k]], base = .045;
                const left = base + attention * (.68 * (!avoiding) * Math.max(0, -Math.sin(e)) + .42 * Math.max(0, -avoid) * 3 + .22 * s.contact * (this.avoidSide < 0));
                const right = base + attention * (.68 * (!avoiding) * Math.max(0, Math.sin(e)) + .42 * Math.max(0, avoid) * 3 + .22 * s.contact * (this.avoidSide > 0));
                this.setTarget(c.pop['PFL3-L'][k], left, dt, .10);
                this.setTarget(c.pop['PFL3-R'][k], right, dt, .10);
            }
            this.setTarget(c.pop.DN[0], clamp(this.mean('PFL3-L') * 2.6), dt, .09);
            this.setTarget(c.pop.DN[1], clamp(this.mean('PFL3-R') * 2.6), dt, .09);
            const heightError = config.altitude - s.clearanceDown;
            this.setTarget(c.pop.ALT[0], clamp(.5 + heightError * .2), dt, .25);
            this.setTarget(c.pop.ALT[1], clamp(.5 - heightError * .2), dt, .25);
            const drive = clamp(this.mean('EPG') * 4, 0, 1);
            let speed = (config.mode === 'walk' ? 3.3 : 5.3) * drive * (1 - clamp(front * .72, 0, .8));
            if (this.odorMode === 'dwell')
                speed *= .4;
            if (s.danger > .3)
                speed *= 1.15;
            if (avoiding) speed = this.recoveryTime ? -2.5 : 0;
            // All yaw control passes through DN activity. No hidden geometric steering.
            this.output = { yawRate: config.motorCoupled ? clamp(3.5 * (this.val('DN', 1) - this.val('DN', 0)), -3, 3) : 0,
                forwardSpeed: config.motorCoupled ? speed : 0, verticalSpeed: config.motorCoupled && config.mode === 'flight' ? 2.7 * (this.val('ALT', 0) - this.val('ALT', 1)) : 0 };
            for (const [i, x] of this.stim) {
                x.remaining -= dt;
                if (x.remaining <= 1e-10)
                    this.stim.delete(i);
            }
            return this.output;
        }
        intervene(ids, kind, amplitude = .65, duration = 2) { for (const id of ids) {
            const i = this.circuit.index[id];
            if (i === undefined)
                throw Error('알 수 없는 뉴런 ' + id);
            if (kind === 'stimulate')
                this.stim.set(i, { amplitude, remaining: duration });
            else if (kind === 'suppress') {
                this.lesions.add(i);
                this.a[i] = 0;
            }
            else if (kind === 'release')
                this.lesions.delete(i);
        } }
        snapshot() { return { a: Array.from(this.a), adapt: Array.from(this.adapt), lesions: [...this.lesions], stim: [...this.stim], rng: this.rng.state, phaseEstimate: this.phaseEstimate, heading: this.heading, confidence: this.confidence, goal: this.goal, clock: this.clock, nextGoal: this.nextGoal, avoidSide: this.avoidSide, prevContact: this.prevContact, odorMode: this.odorMode, turnMemory: this.turnMemory, contactTime: this.contactTime, recoveryTime: this.recoveryTime, output: { ...this.output }, lastCue: this.lastCue, error: this.error, recNorm: this.recNorm, headingBins: this.headingBins || Array(8).fill(0) }; }
        restore(s) {
            const keys = Object.keys(this.snapshot()).sort().join('|');
            if (!s || typeof s !== 'object' || Array.isArray(s) || Object.keys(s).sort().join('|') !== keys)
                throw Error('Unknown neural state fields');
            const num = (v, name, lo = -1e12, hi = 1e12) => {
                if (typeof v !== 'number' || !Number.isFinite(v) || v < lo || v > hi)
                    throw Error('Invalid neural state: ' + name);
            };
            const idx = v => { num(v, 'index', 0, this.circuit.N - 1); if (!Number.isInteger(v)) throw Error('Integer index required'); };
            for (const k of ['a', 'adapt']) {
                if (!Array.isArray(s[k]) || s[k].length !== this.circuit.N) throw Error('Invalid neural shape');
                s[k].forEach(v => num(v, k, 0, 1));
            }
            if (!['explore','heading','chemotaxis','dwell'].includes(s.odorMode)) throw Error('Invalid neural mode');
            if (!s.output || Object.keys(s.output).sort().join('|') !== 'forwardSpeed|verticalSpeed|yawRate') throw Error('Invalid output');
            Object.values(s.output).forEach(v => num(v,'motor',-20,20));
            if (s.lastCue !== null) num(s.lastCue,'lastCue',-Math.PI,Math.PI);
            if (!Array.isArray(s.headingBins) || s.headingBins.length !== 8) throw Error('Invalid heading bins');
            s.headingBins.forEach(v => num(v,'headingBin',0,1));
            if (!Array.isArray(s.lesions) || !Array.isArray(s.stim)) throw Error('Invalid intervention state');
            s.lesions.forEach(idx);
            if (new Set(s.lesions).size !== s.lesions.length) throw Error('Duplicate lesion');
            const used = new Set();
            for (const pair of s.stim) {
                if (!Array.isArray(pair) || pair.length !== 2) throw Error('Invalid stimulus pair');
                const [i,v] = pair; idx(i);
                if (used.has(i)) throw Error('Duplicate stimulus'); used.add(i);
                if (!v || Object.keys(v).sort().join('|') !== 'amplitude|remaining') throw Error('Invalid stimulus');
                num(v.amplitude,'amplitude',-1,1); num(v.remaining,'remaining',1e-12,60);
            }
            num(s.rng,'rng',1,4294967295); if (!Number.isInteger(s.rng)) throw Error('Integer RNG required');
            for (const [k,lo,hi] of [
                ['phaseEstimate',-Math.PI,Math.PI],['heading',-Math.PI,Math.PI],
                ['confidence',0,1+1e-12],['goal',-Math.PI,Math.PI],['clock',0,1e12],['nextGoal',0,1e12],
                ['turnMemory',-2*Math.PI,2*Math.PI],['contactTime',0,1],['recoveryTime',0,.4+1e-12],
                ['error',-Math.PI,Math.PI],['recNorm',0,2+1e-12]]) num(s[k],k,lo,hi);
            if (![-1,1].includes(s.avoidSide) || ![0,1].includes(s.prevContact)) throw Error('Invalid discrete state');
            const { a, adapt, lesions, stim, rng, ...rest } = s;
            this.a = Float64Array.from(a); this.adapt = Float64Array.from(adapt);
            this.lesions = new Set(lesions); this.stim = new Map(F.clone(stim));
            this.rng.state = rng >>> 0; Object.assign(this, F.clone(rest));
        }
    }
    F.NeuralController = NeuralController;
})(globalThis.Fly);

File: flylab/brain.py
"""Python port of A's disclosed hybrid rate controller, not a brain emulation.
The connectome is not changed by adding a physical body. The 72 engineered
nodes are still engineered; no actual PFL3 identifiers are invented.
"""
from __future__ import annotations
import math, re
import numpy as np
from .common import TAU, RNG, clamp, wrap, clone, integer, number

def validate_graph(g):
    if not isinstance(g,dict) or g.get('schema')!='flylab.connectome.v1': raise ValueError('Unsupported connectome schema')
    ns,es=g.get('nodes'),g.get('edges')
    if not isinstance(ns,list) or not 1<=len(ns)<=440 or not isinstance(es,list) or len(es)>100000: raise ValueError('B hybrid backend: 1..440 real IDs, <=100000 edges')
    ids=set()
    for n in ns:
        if not isinstance(n,dict): raise ValueError('Invalid node')
        id=n.get('id'); bid=n.get('bodyId')
        if not isinstance(id,str) or not re.fullmatch(r'hemibrain:\d+',id) or not isinstance(bid,str) or id!='hemibrain:'+bid or id in ids: raise ValueError('Invalid/duplicate ID')
        if n.get('type') not in ('EPG','PENa','PENb','PEG') or n.get('origin')!='anatomical' or n.get('hemisphere') not in ('L','R'): raise ValueError('Unsupported node metadata')
        if not isinstance(n.get('name'),str) or len(n['name'])>200: raise ValueError('Invalid name')
        integer(n.get('indexFix'),'indexFix',1,9); integer(n.get('pbIndex'),'pbIndex',1,9); ids.add(id)
    if not any(n['type']=='EPG' for n in ns): raise ValueError('EPG required by hybrid controller')
    seen=set()
    for e in es:
        if not isinstance(e,dict) or e.get('pre') not in ids or e.get('post') not in ids: raise ValueError('Unknown edge endpoint')
        number(e.get('count'),'count',1e-9,1e9); k=(e['pre'],e['post'])
        if k in seen: raise ValueError('Aggregate ROI rows before import')
        seen.add(k)
    return g

class Circuit:
    def __init__(self,g):
        self.data=clone(validate_graph(g)); self.nodes=clone(g['nodes']); self.real_count=len(self.nodes)
        for n in self.nodes:
            n['angle']=((n['indexFix']-1)%8)*TAU/8
            n['displayName']=f"{n['type']} {n['hemisphere']}{n['pbIndex']} · {n['bodyId'][-4:]}"
        for typ,cnt in [('HD',16),('GOAL',16),('PFL3-L',16),('PFL3-R',16),('DN',2),('SENSE',4),('ALT',2)]:
            for k in range(cnt):
                self.nodes.append(dict(id=f'model:{typ}:{k}',name=f'{typ} {k+1:02d}',displayName=f'{typ} {k+1:02d}',type=typ,origin='model',angle=k/cnt*TAU))
        self.index={n['id']:i for i,n in enumerate(self.nodes)}; self.N=len(self.nodes); self.pop={}
        for i,n in enumerate(self.nodes): self.pop.setdefault(n['type'],[]).append(i)
        # Sparse COO accumulation (no NxN dense matrix); C can replace this backend.
        self.pre=np.array([self.index[e['pre']] for e in g['edges']],dtype=np.int32)
        self.post=np.array([self.index[e['post']] for e in g['edges']],dtype=np.int32)
        weights=np.sqrt([e['count'] for e in g['edges']]); sums=np.bincount(self.post,weights=weights,minlength=self.N)
        self.weight=weights/np.maximum(sums[self.post],1e-30)
    def propagate(self,a,gain=1.): return np.bincount(self.post,weights=self.weight*a[self.pre]*gain,minlength=self.N)
    def catalog(self): return [dict(n,index=i,signal=dict(kind='activation',unit='1',range=[0,1])) for i,n in enumerate(self.nodes)]

class NeuralController:
    def __init__(self,circuit,seed):
        self.circuit=circuit; self.rng=RNG(seed^0x7cab1); self.a=np.full(circuit.N,.05); self.adapt=np.zeros(circuit.N)
        self.lesions=set(); self.stim={}; self.phaseEstimate=0.; self.heading=0.; self.confidence=0.; self.goal=1.
        self.clock=0.; self.nextGoal=9.; self.avoidSide=1; self.prevContact=0; self.odorMode='explore'; self.turnMemory=0.; self.contactTime=0.; self.recoveryTime=0.
        self.output=dict(yawRate=0.,forwardSpeed=0.,verticalSpeed=0.); self.lastCue=None; self.error=0.; self.recNorm=0.; self.headingBins=[0.]*8
    def val(self,typ,k=0): return float(self.a[self.circuit.pop[typ][k]])
    def mean(self,typ): return float(np.mean(self.a[self.circuit.pop[typ]]))
    def target(self,i,value,dt,tau=.13):
        if i in self.lesions: self.a[i]=0; return
        extra=self.stim.get(i,{}).get('amplitude',0)
        self.a[i]+=(clamp(value+extra)-self.a[i])*(1-math.exp(-dt/tau))
    def step(self,s,dt,config):
        # No world, XYZ or ground-truth yaw argument crosses this boundary.
        self.clock+=dt; c=self.circuit; old=self.a.copy(); cx=cy=cs=0.
        for i,x in enumerate(s['panorama']):
            p=i/64*TAU-math.pi; cx+=x*math.cos(p); cy+=x*math.sin(p); cs+=x
        visible=cs>.8; self.lastCue=wrap(-math.atan2(cy,cx)) if visible else None
        self.phaseEstimate=wrap(self.phaseEstimate+s['angularVelocity']*dt)
        if visible: self.phaseEstimate=wrap(self.phaseEstimate+wrap(self.lastCue-self.phaseEstimate)*(1-math.exp(-dt/.24)))
        for i in c.pop['HD']: self.target(i,.05+.90*math.exp(3.4*(math.cos(c.nodes[i]['angle']-self.phaseEstimate)-1)),dt,.07)
        hx=hy=0.
        for i in c.pop['HD']:
            v=max(0,self.a[i]-.04); hx+=v*math.cos(c.nodes[i]['angle']); hy+=v*math.sin(c.nodes[i]['angle'])
        scaffold=math.atan2(hy,hx); rec=c.propagate(old,config['graphGain']); self.recNorm=float(rec.sum()/c.real_count)
        for i in range(c.real_count):
            n=c.nodes[i]; bump=math.exp(3*(math.cos(n['angle']-scaffold)-1)); omega=clamp(s['angularVelocity']/2,-1,1)
            target=.055+.77*bump
            if n['type'] in ('PENa','PENb'): target=.07+.58*bump+.18*max(0,omega*(1 if n['hemisphere']=='R' else -1))
            if n['type']=='PEG': target=.07+.66*bump
            target+=.30*rec[i]-.045; self.adapt[i]+=(old[i]-self.adapt[i])*dt/2; target-=.025*self.adapt[i]
            self.target(i,target,dt,.28 if n['type']=='PEG' else .11)
        bins=np.zeros(8); counts=np.zeros(8)
        for i in c.pop['EPG']:
            k=(c.nodes[i]['indexFix']-1)%8; bins[k]+=max(0,self.a[i]-.025); counts[k]+=1
        bins/=np.maximum(1,counts); angles=np.arange(8)*TAU/8
        ex=float(bins@np.cos(angles)); ey=float(bins@np.sin(angles)); total=float(bins.sum())
        self.confidence=math.hypot(ex,ey)/total if total>.01 else 0.
        if total>.01: self.heading=math.atan2(ey,ex)
        self.headingBins=bins.tolist()
        if not visible and self.confidence>.1: self.phaseEstimate=wrap(self.phaseEstimate+wrap(self.heading-self.phaseEstimate)*dt*.08)
        ranges=s['nearRanges']; dl=max(math.exp(-r/2.2) for r in ranges[:4]); dr=max(math.exp(-r/2.2) for r in ranges[5:]); front=math.exp(-ranges[4]/2.2)
        # Engineered avoidance memory: preserve a turn until proprioceptive
        # rotation and frontal clearance agree that it is safe to walk again.
        # No world coordinates, true yaw, or direct motor bypass enter this path.
        clearance=min(ranges[3:6])
        if self.turnMemory:
            direction=1 if self.turnMemory>0 else -1
            remaining=clamp(abs(self.turnMemory)-direction*s['angularVelocity']*dt,.001,math.tau)
            self.turnMemory=direction*remaining
            if remaining<=.05 and clearance>4.5 and not s['contact']: self.turnMemory=0.
        elif clearance<3.5 or s['contact']:
            if abs(dl-dr)>.08: self.avoidSide=1 if dl>dr else -1
            self.turnMemory=self.avoidSide*1.3
        self.prevContact=s['contact']
        avoiding=bool(self.turnMemory)
        self.recoveryTime=max(0.,self.recoveryTime-dt)
        self.contactTime=clamp(self.contactTime+(dt if s['contact'] and avoiding else -dt*.5),0.,1.)
        if avoiding and not self.recoveryTime and self.contactTime>.3:
            self.recoveryTime=.4
            self.contactTime=0.

        for i,v in enumerate([dl,dr,*s['odor']]): self.target(c.pop['SENSE'][i],v,dt,.08 if i<2 else .15)
        if config['task']=='heading': self.goal=config['goalAngle']; self.odorMode='heading'
        else:
            if self.clock>self.nextGoal:
                self.goal=wrap(self.goal+(.5+self.rng.next())*self.rng.signed()*2); self.nextGoal=self.clock+7+self.rng.next()*7
            if max(s['odor'])>.10:
                self.goal=wrap(self.heading+clamp((self.val('SENSE',3)-self.val('SENSE',2))*9,-1,1)+(.50 if s['odorChange']<-.02 else 0)); self.odorMode='chemotaxis'
            else: self.odorMode='explore'
            if max(s['odor'])>.88: self.goal=wrap(self.heading+.7); self.odorMode='dwell'
        for i in c.pop['GOAL']: self.target(i,.02+.94*math.exp(3.3*(math.cos(c.nodes[i]['angle']-self.goal)-1)),dt,.2)
        gx=sum(self.a[i]*math.cos(c.nodes[i]['angle']) for i in c.pop['GOAL']); gy=sum(self.a[i]*math.sin(c.nodes[i]['angle']) for i in c.pop['GOAL'])
        self.error=wrap(math.atan2(gy,gx)-self.heading)
        avoid=clamp(self.val('SENSE',0)-self.val('SENSE',1)+max(0,s['danger']-.25)*self.avoidSide,-1,1)
        if avoiding: avoid=.85*(1 if self.turnMemory>0 else -1)
        for k in range(16):
            att=1. if avoiding else .30+.70*self.a[c.pop['GOAL'][k]]
            left=.045+att*(.68*(not avoiding)*max(0,-math.sin(self.error))+.42*max(0,-avoid)*3+.22*s['contact']*(self.avoidSide<0))
            right=.045+att*(.68*(not avoiding)*max(0,math.sin(self.error))+.42*max(0,avoid)*3+.22*s['contact']*(self.avoidSide>0))
            self.target(c.pop['PFL3-L'][k],left,dt,.10); self.target(c.pop['PFL3-R'][k],right,dt,.10)
        self.target(c.pop['DN'][0],clamp(self.mean('PFL3-L')*2.6),dt,.09); self.target(c.pop['DN'][1],clamp(self.mean('PFL3-R')*2.6),dt,.09)
        he=config['altitude']-s['clearanceDown']
        self.target(c.pop['ALT'][0],clamp(.5+he*.2),dt,.25); self.target(c.pop['ALT'][1],clamp(.5-he*.2),dt,.25)
        speed=(3.3 if config['mode']=='walk' else 5.3)*clamp(self.mean('EPG')*4)*(1-clamp(front*.72,0,.8))
        if self.odorMode=='dwell': speed*=.4
        if s['danger']>.3: speed*=1.15
        if avoiding: speed=-2.5 if self.recoveryTime else 0.  # back away, then pivot via CPG
        coupled=config['motorCoupled']
        self.output=dict(yawRate=clamp(3.5*(self.val('DN',1)-self.val('DN',0)),-3,3) if coupled else 0.,forwardSpeed=float(speed) if coupled else 0.,verticalSpeed=2.7*(self.val('ALT',0)-self.val('ALT',1)) if coupled and config['mode']=='flight' else 0.)
        for i in list(self.stim):
            self.stim[i]['remaining']-=dt
            if self.stim[i]['remaining']<=1e-10: del self.stim[i]
        return dict(self.output)
    def intervene(self,ids,kind,amplitude=.65,duration=2):
        for id in ids:
            i=self.circuit.index[id]
            if kind=='stimulate': self.stim[i]=dict(amplitude=amplitude,remaining=duration)
            elif kind=='suppress': self.lesions.add(i); self.a[i]=0
            elif kind=='release': self.lesions.discard(i)
    def snapshot(self):
        scalar=['phaseEstimate','heading','confidence','goal','clock','nextGoal','avoidSide','prevContact','odorMode','turnMemory','contactTime','recoveryTime','output','lastCue','error','recNorm','headingBins']
        return dict(a=self.a.tolist(),adapt=self.adapt.tolist(),lesions=sorted(self.lesions),stim=[[i,clone(x)] for i,x in self.stim.items()],rng=self.rng.state,**{k:clone(getattr(self,k)) for k in scalar})
    def restore(self,s):
        """Validate the entire state before committing it (atomic on rejection).

        fix_2 changes validation only, not the forward dynamics or checkpoint
        schema. Valid v0.2.1 snapshots, including active recovery, remain valid.
        """
        if not isinstance(s,dict) or set(s)!=set(self.snapshot()):
            raise ValueError('Unknown neural state fields')
        if s['odorMode'] not in ('explore','heading','chemotaxis','dwell'):
            raise ValueError('Invalid neural mode')
        if not isinstance(s['output'],dict) or set(s['output'])!={'yawRate','forwardSpeed','verticalSpeed'}:
            raise ValueError('Invalid motor command state')
        for v in s['output'].values(): number(v,'motor',-20,20)
        if s['lastCue'] is not None: number(s['lastCue'],'lastCue',-math.pi,math.pi)
        if not isinstance(s['headingBins'],list) or len(s['headingBins'])!=8:
            raise ValueError('Invalid heading bins')
        for v in s['headingBins']: number(v,'headingBin',0,1)
        arrays={}
        for k in ('a','adapt'):
            if not isinstance(s[k],list): raise ValueError('Neural array must be a list')
            for v in s[k]: number(v,k,0,1)
            arr=np.asarray(s[k],dtype=float)
            if arr.shape!=(self.circuit.N,) or not np.isfinite(arr).all() or (arr<0).any() or (arr>1).any():
                raise ValueError('Invalid neural state')
            arrays[k]=arr.copy()
        if not isinstance(s['lesions'],list) or not isinstance(s['stim'],list):
            raise ValueError('Invalid intervention state')
        lesions=[]
        for i in s['lesions']: lesions.append(integer(i,'lesion',0,self.circuit.N-1))
        if len(set(lesions))!=len(lesions): raise ValueError('Duplicate lesion')
        stimuli={}
        for pair in s['stim']:
            if not isinstance(pair,list) or len(pair)!=2: raise ValueError('Invalid stimulus pair')
            i,v=pair; integer(i,'stim',0,self.circuit.N-1)
            if i in stimuli: raise ValueError('Duplicate stimulus')
            if not isinstance(v,dict) or set(v)!={'amplitude','remaining'}:
                raise ValueError('Invalid stimulus state')
            number(v['amplitude'],'amplitude',-1,1)
            number(v['remaining'],'remaining',1e-12,60)
            stimuli[i]=clone(v)
        rng=integer(s['rng'],'rng',1,2**32-1)
        # State domains follow step(); a tiny time tolerance accommodates JSON
        # floating-point round trips, not seconds-long extra recovery periods.
        for k,lo,hi in (
            ('phaseEstimate',-math.pi,math.pi),('heading',-math.pi,math.pi),
            ('confidence',0,1+1e-12),('goal',-math.pi,math.pi),
            ('clock',0,1e12),('nextGoal',0,1e12),
            ('turnMemory',-TAU,TAU),('contactTime',0,1),
            ('recoveryTime',0,.4+1e-12),('error',-math.pi,math.pi),('recNorm',0,2+1e-12)):
            number(s[k],k,lo,hi)
        number(s['avoidSide'],'avoidSide',-1,1)
        if s['avoidSide'] not in (-1,1): raise ValueError('avoidSide must be -1 or 1')
        number(s['prevContact'],'prevContact',0,1)
        if s['prevContact'] not in (0,1): raise ValueError('prevContact must be 0 or 1')
        # No self mutation above this point, including RNG and activity buffers.
        self.a=arrays['a']; self.adapt=arrays['adapt']; self.lesions=set(lesions)
        self.stim=stimuli; self.rng.state=rng
        for k,v in s.items():
            if k not in ('a','adapt','lesions','stim','rng'): setattr(self,k,clone(v))
